app/models/railway_carriage.rb

```diff
@@ -0,0 +1,13 @@
+class RailwayCarriage < ApplicationRecord
+  validates :number, presence: true
+  validates :number, uniqueness: { scope: :train_id }
+  belongs_to :train
+
+  before_validation :set_number
+
+  private
+
+  def set_number
+    self.number = train.railway_carriages.size + 1
+  end
```

Comment:
Are you sure that it works right?) Check it please, because when you have 2 carriages for example, with number 1 and 2, and when you update any of them, it will get number 3. And when you update another one, you have troubles with validation, because they both will be number 3.